### ISSUE TITLE
Possibly incorrect aggregation?

### ISSUE THREAD

**Issue by max-sixty** [re-actors/alls-green#23]
Thanks for building this!

One issue today though — am I reading this wrong? It seems like the two failures are allowed, yet the job aggregates this as a failure?

https://github.com/PRQL/prql/actions/runs/5942757316

```
# ❌ Some of the required to succeed jobs failed 😢😢😢

🛈 Some of the allowed to fail jobs did not succeed.
🛈 Some of the allowed to be skipped jobs did not succeed.
📝 Job statuses:
📝 build-web → ✓ success [required to succeed or be skipped]
📝 check-links-book → ✓ success [required to succeed or be skipped]
📝 check-links-markdown → ❌ failure [allowed to fail]
📝 lint-megalinter → ✓ success [required to succeed or be skipped]
📝 nightly → ❌ failure [allowed to fail]
📝 publish-web → ⬜ skipped [required to succeed or be skipped]
📝 test-dotnet → ✓ success [required to succeed or be skipped]
📝 test-elixir → ✓ success [required to succeed or be skipped]
📝 test-java → ✓ success [required to succeed or be skipped]
📝 test-js → ✓ success [required to succeed or be skipped]
📝 test-lib → ✓ success [required to succeed or be skipped]
📝 test-php → ✓ success [required to succeed or be skipped]
📝 test-python → ✓ success [required to succeed or be skipped]
📝 test-rust → ✓ success [required to succeed or be skipped]
📝 test-rust-main → ✓ success [required to succeed or be skipped]

```

**Comment by webknjaz** [comment]
Looks like an input validation issue. The `allowed-skips` input accepts either a comma-separated or a JSON list. But you're passing a mapping object here: https://github.com/PRQL/prql/actions/runs/5942757316/workflow#L477 (serialized as a string; the value is visible at https://github.com/PRQL/prql/actions/runs/5942757316/job/16118911482#step:2:4).

We'll need to error out if https://github.com/re-actors/alls-green/blob/223e4bb/src/normalize_needed_jobs_status.py#L45 returns a ~~dict~~non-list.

P.S. In your case, if you want to allow everything to be skipped, you'll have to pre-process the jobs dict before passing it as an action input, transforming it into an acceptable list format.

**Comment by max-sixty** [comment]
Thanks @webknjaz , that makes sense, appreciate you looking.

One question though — why does the report suggest that all the jobs are either allowed to fail or be skipped? Or am I reading it incorrectly?

**Comment by webknjaz** [comment]
> i.e. the report seems to interpret the input correctly, even though it's incorrectly formatted?

Fair enough. Upon closer inspection, I think that there might be an actual bug with `allowed-failures` @ https://github.com/re-actors/alls-green/blob/223e4bb7a751b91f43eda76992bcfbf23b8b0302/src/normalize_needed_jobs_status.py#L162C12-L162C46.

**Comment by max-sixty** [comment]
I think this might be a case of it not working, after having made the suggested change:

https://github.com/PRQL/prql/actions/runs/5967503062/job/16189922489?pr=3391

```

📝 Job statuses:
📝 build-web → ✓ success [required to succeed or be skipped]
📝 check-links-book → ✓ success [required to succeed or be skipped]
📝 check-links-markdown → ✓ success [allowed to fail]
📝 lint-megalinter → ✓ success [required to succeed or be skipped]
📝 nightly → ❌ failure [allowed to fail]
📝 publish-web → ⬜ skipped [required to succeed or be skipped]
📝 test-dotnet → ✓ success [required to succeed or be skipped]
📝 test-elixir → ✓ success [required to succeed or be skipped]
📝 test-java → ✓ success [required to succeed or be skipped]
📝 test-js → ✓ success [required to succeed or be skipped]
📝 test-lib → ✓ success [required to succeed or be skipped]
📝 test-php → ✓ success [required to succeed or be skipped]
📝 test-python → ✓ success [required to succeed or be skipped]
📝 test-rust → ✓ success [required to succeed or be skipped]
📝 test-rust-main → ✓ success [required to succeed or be skipped]
```

**Comment by webknjaz** [comment]
I reproduced this locally with your inputs and I think I have a patch. I'm going to push a commit for you to test, but will probably not make a release just yet. Instead, I'll attempt adding some more comprehensive testing before doing that.

**Comment by webknjaz** [comment]
@max-sixty Try pinning the action to commit `cf9edfcf932a0ed6b431433fa183829c68b30e3f`. Though, the smoke-test in the PR is failing. So I'll need to inspect the GHA workflow to see if the test is incorrect or the fix is incomplete. Anyway, I already have some thoughts of refactoring, but that can't be done w/o proper testing, especially now that some major projects depend on the action and I don't want to break anybody...

**Comment by max-sixty** [comment]
Great, I put that in, I'll keep an eye on it!

**Comment by krokofant** [comment]
@webknjaz Isn't this just an error with the log?

https://github.com/re-actors/alls-green/blob/72891fe48919b57e202f981aa1edc5e892c9932a/src/normalize_needed_jobs_status.py#L99C1-L106C10

Shouldn't this be
```
    if jobs_allowed_to_be_skipped and allowed_to_be_skipped_jobs_succeeded:
        markdown_summary_lines += {
            '🛈 All of the allowed to be skipped dependency jobs succeeded.',
        }
    elif jobs_allowed_to_be_skipped:
        markdown_summary_lines += {
            '🛈 Some of the allowed to be skipped jobs did not succeed.',
        }
```

The elif checked against the `jobs_allowed_to_fail` instead of `jobs_allowed_to_be_skipped`

**Comment by krokofant** [comment]
Yes, and I misunderstood the fix. I started looking at the tests and found it weird that the tests failed at all.

I didn't find that your conversation made it really clear what the issue was since those old CI runs are long gone. But now I understand the change you made, and why those tests are failing. I didn't like the idea to use this while the tests were failing.

The behavior change in the commit is that skipped jobs in allowed-failures will fail the check.

Old 1.2.2 behavior

|                   | success | skipped | failure |
| ----------------- | :-----: | :-----: | :-----: |
| no exceptions     |   ✅    |   ❌    |   ❌    |
| allow skip        |   ✅    |   ✅    |   ❌    |
| allow fail        |   ✅    |   ✅    |   ✅    |
| allow fail & skip |   ✅    |   ✅    |   ✅    |

New cf9edfcf932a0ed6b431433fa183829c68b30e3f behavior

|                   | success | skipped | failure |
| ----------------- | :-----: | :-----: | :-----: |
| no exceptions     |   ✅    |   ❌    |   ❌    |
| allow skip        |   ✅    |   ✅    |   ❌    |
| allow fail        |   ✅    |   ❌    |   ✅    |
| allow fail & skip |   ✅    |   ✅    |   ✅    |

So the tests needs to be updated to remove the skipped jobs from the allow-failure jobs.

**Comment by krokofant** [comment]
The tests would look like this instead
```diff
diff --git a/.github/workflows/self-smoke-test-action.yml b/.github/workflows/self-smoke-test-action.yml
--- a/.github/workflows/self-smoke-test-action.yml
+++ b/.github/workflows/self-smoke-test-action.yml
@@ -105,7 +105,6 @@ jobs:
     needs:
     - failing-job
     - succeeding-job
-    - skipped-job
     if: always()
     runs-on: ubuntu-latest
     steps:
@@ -117,7 +116,7 @@ jobs:
       continue-on-error: true
       with:
         allowed-failures: >-
-          ["failing-job", "skipped-job"]
+          ["failing-job"]
         jobs: ${{ toJSON(needs) }}
     - name: Fail on unexpected outcome (assertion step)
       run: exit 1
@@ -129,7 +128,6 @@ jobs:
     needs:
     - failing-job
     - succeeding-job
-    - skipped-job
     if: always()
     runs-on: ubuntu-latest
     steps:
@@ -140,7 +138,7 @@ jobs:
       id: check
       continue-on-error: true
       with:
-        allowed-failures: failing-job, skipped-job
+        allowed-failures: failing-job, succeeding-job
         jobs: ${{ toJSON(needs) }}
     - name: Fail on unexpected outcome (assertion step)
       run: exit 1
```

Could this be merged and release as v2? 😅

**Comment by webknjaz** [comment]
I don't know about v2 (the public API hasn't changed, so likely another v1.x is in order), but I've started adding actual tests to have observable behavior changes while bugfixing. This will help us validate correctness better and finally complete #24 + add more cosmetic changes I've got lined up.

**Comment by webknjaz** [comment]
@krokofant FYI, I ended up picking your first suggestion into `unstable/v1` as https://github.com/re-actors/alls-green/commit/2410c4cfd1739b8daac80b163f67b3bc7e82fa8a. Separately from #24. It's got some rewrite and pytest based testing that I'm working on meanwhile.

**Comment by webknjaz** [comment]
Looks like this has actually been fixed by the latest refactoring commits on the default repo branch.

**Comment by webknjaz** [comment]
Apologies for the delay: https://github.com/re-actors/alls-green/releases/tag/v1.3.0 / https://github.com/re-actors/alls-green/discussions/41.
